scipy/interpolate/_aaa.py

```diff
@@ -0,0 +1,404 @@
+           -1.99999988e+00+0.j        , -1.00000000e+00+0.j        ,
+            5.85842812e-17+0.j        ,  4.77485458e+00-3.06919376j,
+            4.77485458e+00+3.06919376j,  5.29095868e+00-0.97373072j,
+            5.29095868e+00+0.97373072j])
+    >>> r.residues()[order]
+    array([ 0.03658074 +0.j        , -0.16915426 -0.j        ,
+            0.49999915 +0.j        , -1.         +0.j        ,
+            1.         +0.j        , -0.81132013 -2.30193429j,
+           -0.81132013 +2.30193429j,  0.87326839+10.70148546j,
+            0.87326839-10.70148546j])
+
+    For the second example, we call `AAA` with a spiral of 1000 points that wind 7.5
+    times around the origin in the complex plane.
+
+    >>> z = np.exp(np.linspace(-0.5, 0.5 + 15j*np.pi, 1000))
+    >>> r = AAA(z, np.tan(np.pi*z/2), rtol=1e-13)
+
+    We see that AAA takes 12 steps to converge with the following errors:
+
+    >>> r.errors.size
+    12
+    >>> r.errors
+    array([2.49261500e+01, 4.28045609e+01, 1.71346935e+01, 8.65055336e-02,
+           1.27106444e-02, 9.90889874e-04, 5.86910543e-05, 1.28735561e-06,
+           3.57007424e-08, 6.37007837e-10, 1.67103357e-11, 1.17112299e-13])
+
+    We can also plot the computed poles:
+
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(z.real, z.imag, '.', markersize=2, label="Sample points")
+    >>> ax.plot(r.poles().real, r.poles().imag, '.', markersize=5,
+    ...         label="Computed poles")
+    >>> ax.set(xlim=[-3.5, 3.5], ylim=[-3.5, 3.5], aspect="equal")
+    >>> ax.legend()
+    >>> plt.show()
+    """
+    def __init__(self, z, f, *, rtol=None, max_terms=100):
+        # input validation
+        z = np.asarray(z)
+        f = np.asarray(f)
+
+        if z.ndim != 1 or f.ndim != 1:
+            raise ValueError("`f` and `z` must be 1-D.")
+
+        if f.size != z.size:
+            raise ValueError("`f` and `z` must be the same size.")
+
+        # Remove infinite or NaN function values and repeated entries
+        to_keep = (np.isfinite(f)) & (~np.isnan(f))
+        f = f[to_keep]
+        z = z[to_keep]
+        z, uni = np.unique(z, return_index=True)
+        f = f[uni]
+
+        self._compute_AAA(f, z, rtol, max_terms)
+
+        # only compute once
+        self._poles = None
+        self._residues = None
+        self._roots = None
+
+    def _compute_AAA(self, f, z, rtol, max_terms):
+        # Initialization for AAA iteration
+        M = np.size(z)
+        mask = np.ones(M, dtype=np.bool_)
+        dtype = np.result_type(z, f, 1.0)
+        rtol = np.finfo(dtype).eps**0.75 if rtol is None else rtol
+        atol = rtol * np.linalg.norm(f, ord=np.inf)
+        zj = np.empty(max_terms, dtype=dtype)
+        fj = np.empty(max_terms, dtype=dtype)
+        # Cauchy matrix
+        C = np.empty((M, max_terms), dtype=dtype)
+        # Loewner matrix
+        A = np.empty((M, max_terms), dtype=dtype)
+        errors = np.empty(max_terms)
```

Comment:
I did consider this but the error is the output of a norm so is always real. This probably be the right size floating type rather than possibly complex if `dtype` is complex

Reply:
Oops!
```suggestion
        errors = np.empty(max_terms, dtype=A.real.dtype)
```